app/uk/gov/hmrc/helptosavefrontend/controllers/RegisterController.scala

```diff
@@ -18,60 +18,58 @@ package uk.gov.hmrc.helptosavefrontend.controllers
 
 import javax.inject.Singleton
 
-import cats.data.{EitherT, OptionT, ValidatedNel}
+import cats.data.{EitherT, ValidatedNel}
 import cats.instances.future._
 import cats.syntax.either._
 import com.google.inject.Inject
-import play.api.Logger
 import play.api.i18n.{I18nSupport, MessagesApi}
-import play.api.mvc.Action
-import uk.gov.hmrc.helptosavefrontend.auth.HtsCompositePageVisibilityPredicate.twoFactorURI
+import play.api.mvc.{Action, AnyContent}
+import play.api.{Application, Logger}
 import uk.gov.hmrc.helptosavefrontend.connectors.NSIConnector.{SubmissionFailure, SubmissionResult, SubmissionSuccess}
 import uk.gov.hmrc.helptosavefrontend.connectors._
 import uk.gov.hmrc.helptosavefrontend.models.{HTSSession, NSIUserInfo, UserInfo}
-import uk.gov.hmrc.helptosavefrontend.models.UserInfo
-import uk.gov.hmrc.helptosavefrontend.services.userinfo.UserInfoService
+import uk.gov.hmrc.helptosavefrontend.services.HelpToSaveService
 import uk.gov.hmrc.helptosavefrontend.util.Result
 import uk.gov.hmrc.helptosavefrontend.views
-import uk.gov.hmrc.play.frontend.auth.AuthContext
-import uk.gov.hmrc.play.frontend.auth.connectors.domain.Accounts
 import uk.gov.hmrc.play.http.HeaderCarrier
 
 import scala.concurrent.{ExecutionContext, Future}
 
 @Singleton
-class RegisterController @Inject()(val sessionCacheConnector: SessionCacheConnector,
-                                   val messagesApi: MessagesApi,
-                                   eligibilityConnector: EligibilityConnector,
-                                   citizenDetailsConnector: CitizenDetailsConnector,
-                                   nSAndIConnector: NSIConnector) extends HelpToSaveController with I18nSupport {
-
-  val userInfoService = new UserInfoService(authConnector, citizenDetailsConnector)
-
-  def declaration =
-    AuthorisedHtsUserAction { implicit authContext ⇒
-      implicit request ⇒
-        validateUser(authContext).fold(
+class RegisterController @Inject()(override val messagesApi: MessagesApi,
+                                   htsService: HelpToSaveService,
+                                   nSAndIConnector: NSIConnector,
+                                   val sessionCacheConnector: SessionCacheConnector,
+                                   implicit val app: Application)
+  extends HelpToSaveAuth(app) with I18nSupport {
+
+  def declaration: Action[AnyContent] = authorisedForHtsWithEnrolments {
+    implicit request ⇒
+      implicit userUrlWithNino ⇒
+        val userInfo = for {
+          userUrlWithNino <- EitherT.fromOption[Future](userUrlWithNino, "could not retrieve either userDetailsUrl or NINO from auth")
+          result <- checkEligibility(userUrlWithNino.path, userUrlWithNino.nino)
```

Comment:
Do we want to put the session cache code in this for comprehension? e.g.
```
for {
...
result <- checkEligibility(userUrlWithNino.path, userUrlWithNino.nino)
_       <- f(result)
} yield result

...
def f(userDetails: Option[UserInfo]): EitherT[Future,String,CacheMap] = {
  EitherT.fromOption[Future](userDetails, "No user details to write to key-store")
    .flatMapF(details ⇒
      sessionCacheConnector.put(HTSSession(Some(details)))
        .map[Either[String,CacheMap]](Right(_))
        .recover{ case e ⇒ Left(s"Could not write to key-store: ${e.getMessage}")}
    )
}
```
This way we make sure we have stored the information in key-store before proceeding